### ISSUE TITLE
PyTorch model code release 

### ISSUE THREAD

**Issue by PeterZhizhin** [graphdeco-inria/gaussian-splatting#5]
Thank you very much for releasing the source code!

I see that the model is fully implemented in CUDA. However, for experimentation purposes, pure PyTorch implementation would be better.

Is there a plan to release the PyTorch model implementation?

Thank you!

**Comment by grgkopanas** [comment]
Hi Peter,

The released code is provided as is, unfortunately, we do not have plans for a full pytorch release. There are some parts of the code in pytorch already, like the optimization loop and there are some flags inside the code that allow some of the operations like the computation of the covariance matrix in 2D to be done in pytorch.

Any pytorch code you write that involves the properties of the point cloud before the actual rendering function will work properly with auto-diff but if you would want to change the actual rendering you would need to do it in cuda.

Best,
George

**Comment by kwea123** [comment]
> for experimentation purposes, pure PyTorch implementation would be better.

CUDA is used to perform some operations a lot faster than pytorch, if you want them in pytorch, you can just rewrite as for loops, but that doesn't make sense, it will just make things much slower.

You cannot get both "simplicity for debugging" and "speed".

**Comment by jianhengLiu** [comment]
> Hi Peter,
> 
> The released code is provided as is, unfortunately, we do not have plans for a full pytorch release. There are some parts of the code in pytorch already, like the optimization loop and there are some flags inside the code that allow some of the operations like the computation of the covariance matrix in 2D to be done in pytorch.
> 
> Any pytorch code you write that involves the properties of the point cloud before the actual rendering function will work properly with auto-diff, or even after rendering if you want to do something with the loss function, but if you would want to change the actual rendering you would need to do it in cuda.
> 
> Best, George

Excellent works!
I am also trying to modify codes to add some features, while I am not familiar with the development chain of CUDA.
It takes lots of effort to validate or debug, like checking whether gradients are correct or not.

Is there any suggestion or tool recommended?

Thanks a lot!
